src/hotspot/share/services/mallocTracker.cpp

```diff
@@ -70,6 +70,23 @@ size_t MallocMemorySnapshot::total_arena() const {
   return amount;
 }
 
+size_t MallocMemorySnapshot::thread_count() const {
+  if (ThreadStackTracker::track_as_vm()) {
+    return _thread_count;
+  } else {
+    assert(_thread_count == 0, "_thread_count can not be used if ThreadStackTracker::track_as_vm() == false");
+    return by_type(mtThreadStack)->malloc_count();
+  }
+}
+
+void MallocMemorySnapshot::snapshot_thread_count() {
+  if (ThreadStackTracker::track_as_vm()) {
+    _thread_count = ThreadStackTracker::thread_count();
+  } else {
+    // Thread count will be reported as malloc count for mtThreadStack.
+  }
+}
+
```

Comment:
I would prefer more explicit code here with respect to `_thread_count`:

```
void MallocMemorySnapshot::snapshot_thread_count() {
  if (ThreadStackTracker::track_as_vm()) {
    _thread_count = ThreadStackTracker::thread_count();
  } else {
    _thread_count = thread_count();
  }
}
```

But if we do that, the we should probably remove:

`    assert(_thread_count == 0, "_thread_count can not be used if ThreadStackTracker::track_as_vm() == false");
`

from `MallocMemorySnapshot::thread_count()`

Not going to insist on this change though.

Reply:
Did anyone ran this through Mach5 tests? At least tiers 1 through 5?

Reply:
Well in case you are not insisting I would like to leave this code as is :) 
Thank you for the review